service/gsheet_bot.py

```diff
@@ -0,0 +1,210 @@
+import gspread
+import pandas as pd
+from google.oauth2.service_account import Credentials
+import os
+from config import Config
+from discord.utils import get
+from logger import Logger
+
+config = Config()
+
+# Google Sheet Configurations
+CREDENTIALS_FILE = os.getenv('GBOT_GOOGLE_SERVICE_ACCOUNT', 'config/services_account.json')
+
+# System Configurations
+GBOT_MINIMUM_VOLUME_THRESHOLD = int(os.getenv('GBOT_MINIMUM_VOLUME_THRESHOLD', 30000))
+
+logger_mod = Logger("GSheet Bot")
+logger = logger_mod.get_logger()
+
+# Global variable for expected headers
+expected_headers = ['uuid', 'discord_id', 'volume']
+
+free_vip_role = 'FREE_VIP'
+
+
+def get_worksheet(sheet_id):
+    client = init_gspread_client()
+    sheet = client.open_by_key(sheet_id)
+    return sheet.get_worksheet(0)
+
+
+def init_gspread_client():
+    scope = [
+        'https://www.googleapis.com/auth/spreadsheets',
+        'https://www.googleapis.com/auth/drive'
+    ]
+    credentials = Credentials.from_service_account_file(CREDENTIALS_FILE, scopes=scope)
+    return gspread.authorize(credentials)
+
+
+def get_all_user(worksheet):
+    records = worksheet.get_all_records(expected_headers=expected_headers, head=2)
+    df = pd.DataFrame(records)
+
+    # Ensure all volume values are floats
+    df['volume'] = df['volume'].astype(float)
+
+    # Summarize volumes by UUID
+    grouped_df = df.groupby('uuid', as_index=False).agg({
+        'discord_id': 'first',
+        'volume': 'sum'
+    })
+
+    grouped_df['status'] = grouped_df['volume'].apply(lambda x: 'pass' if x >= GBOT_MINIMUM_VOLUME_THRESHOLD else 'fail')
+
+    return grouped_df
+
+
+async def batch_assign_free_vip(dbcon, guild):
+    worksheet = get_worksheet(config.GBOT_GSHEET_ID)
+    df = get_all_user(worksheet)
+
+    if df.empty:
+        logger.info("No data found. Skipping update.")
+        return True
+
+    df_pass = df[df['status'] == 'pass']
+
+    if df_pass.empty:
+        logger.info("No pass user to be updated.")
+        return True
+
+    update_scripts = []
+    insert_scripts = []
+    free_vip_discord_ids = []
+
+    for _, row in df_pass.iterrows():
+        uuid = row['uuid']
+        discord_id = row['discord_id']
+        volume = row['volume']
+        status = row['status']
+
+        if not pd.notna(uuid) or not pd.notna(discord_id):
+            logger.info("Empty UUID or Discord ID found. Skipping entry.")
+            continue
+
+        free_vip_discord_ids.append(discord_id)
+
+        # Group all insert/update trade volume table
+        if dbcon.fetch_user_trade_volume_by_discord_id(discord_id):
```

Comment:
this doesnt seem correct, i will need to call this function 1000 times, my db pool connection will be fully occupied

Reply:
can we do something like this?
`IF EXISTS(SELECT from table WHERE c=3)
BEGIN
UPDATE table SET a=1, b=2 WHERE c=3
END

ELSE
BEGIN
INSERT into table(a,b,c) VALUES(1,2,3)
END`

if not, do you have other good way?